[wip] service_bots: Add new event trigger.

PieterCK opened the pull request:
This PR is pending on: #35250

---

This PR adds a new Service trigger: 
- `"all_received"`  which activates whenever the bot receives any message.

### Custom migration testing plan
The migration in `migrations: Fix service bot with not triggers.` updates all service bots that have an empty triggers list. I manually tested the custom migration by following these steps:
<details>

<summary>
test plan:
</summary>

1. Rebase and reset to main branch
2. Create 2 Service bots `outgoing1` and `embedded1`
3. Rebase to the commits in this PR and migrate to `07xx_service_triggers.py`
4. Inspect the `triggers` fields of the service bots created in step 2, they should be empty:
  ```
  # db shell
  In [1]: from zerver.models.bots import Service
  
  In [2]: Service.objects.get(user_profile=24).triggers
  Out[2]: []
  
  In [3]: Service.objects.get(user_profile=25).triggers
  Out[3]: []

  ```
6. Migrate to `07xx_fix_bot_services_with_empty_triggers.py`
7. Inspect the updated `triggers` fields: 
  ```
  # db shell
  In [4]: Service.objects.get(user_profile=25).triggers
  Out[4]: ['all_mentions', 'dm_received']
  
  In [5]: Service.objects.get(user_profile=24).triggers
  Out[5]: ['all_mentions', 'dm_received']
  ```

I also printed out the query result in the custom migration and it only showed 3 results, 2 I added manually and 1 is AARON's outgoing bot:
```
Running migrations:
  Applying zerver.0711_fix_bot_services_with_empty_triggers...<QuerySet [<Service: Service object (1)>, <Service: Service object (2)>, <Service: Service object (3)>]>
 OK
```
Running the same migration again results in no query:
```
Running migrations:
  Applying zerver.0711_fix_bot_services_with_empty_triggers...<QuerySet []>
 OK 
```

</details>

**Screenshots and screen captures:**
(updated June 17)
| Add new a bot | Edit bot |
|--------|--------|
| Outgoing webhook bot ![image](https://github.com/user-attachments/assets/25d69e19-1e65-4bc2-afa8-0f2cc17ff40f) | ![image](https://github.com/user-attachments/assets/79d446a6-903e-4509-ad63-21a73dda1ab3)|
| Embedded bot ![image](https://github.com/user-attachments/assets/0e471a8f-7b1e-4234-b71b-3a727a7bc0dd) | (we currently don't support editing embedded bots service) |
|no triggers selected error ![image](https://github.com/user-attachments/assets/e2754e3f-0c0f-47bf-a1f3-ea660dfc4c57) | ![image](https://github.com/user-attachments/assets/e707fc2a-1243-4b30-9d45-6112f2e94c68) |

---

(some UI strings might be outdated, but there has been no change to the code underneath. June 17)
Adding new outgoing webhook bot:

[Screencast from 2025-06-13 12-09-13.webm](https://github.com/user-attachments/assets/f5d2dc27-40f2-40d8-91c5-567c726a0724)

Editing the bot:

[Screencast from 2025-06-13 12-11-07.webm](https://github.com/user-attachments/assets/3587d9a0-2622-4cd9-9aa8-4f1088e1c482)

New bot has the correct triggers:

[Screencast from 2025-06-13 13-49-16.webm](https://github.com/user-attachments/assets/d31d3507-00eb-40aa-bf8c-2b9145347f34)

I did some manual tests to see how the different triggers work together, but I think tests in `test_service_bot_system` do a better job at showing that than screencap or screenshots. Especially `test_multiple_trigger_types`.

---

(updated June 17)
| API Doc `/api/get-events` | Help doc `help/bot-overview`|
|--------|---------|
| (updated)![image](https://github.com/user-attachments/assets/f91a7770-44dc-420c-a818-6c94e965fc98) | (added a new section)![image](https://github.com/user-attachments/assets/b8d89149-fc86-4da7-922f-b9ad4bde4fcb) |

| API doc `api/outgoing-webhook` (updated July 2) |
|--------|
| (updated) ![image](https://github.com/user-attachments/assets/3f1559bb-8f17-4e8f-9d7b-3c2482ee4cb9) | 

Feature request discussion: [CZO](https://chat.zulip.org/#narrow/channel/2-general/topic/Feature.20feasibility.20check.20-.20expanding.20outgoing.20webhooks/near/2163683)
API discussion: [#api design > Adding service bot trigger types. #34524](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Adding.20service.20bot.20trigger.20types.2E.20.2334524/with/2169743)
design discussion: [#design > Service bot triggers UI](https://chat.zulip.org/#narrow/channel/101-design/topic/Service.20bot.20triggers.20UI/with/2181554) & [#design > Nested checkboxes in settings menu.](https://chat.zulip.org/#narrow/channel/101-design/topic/Nested.20checkboxes.20in.20settings.20menu.2E/with/2189802)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK committed:
message_send: Refactor service bots trigger system.

This adds a new field to the Service table called
"triggers", which specifes the bots' list of triggers. This
also splits out existing triggers (mention and direct message) into
their own trigger types. All existing Service bots should have
["all_mentions", "dm_received"] as their Service.triggers field.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -513,6 +519,7 @@ def get_service_bot_events(
     mentioned_user_ids: set[int],
     active_user_ids: set[int],
     recipient_type: int,
+    recipient_type_id: int,
```
Might be better to just pass the `Recipient` object from the caller?

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -535,35 +542,54 @@ def maybe_add_event(user_profile_id: int, bot_type: int) -> None:
             return
 
         is_stream = recipient_type == Recipient.STREAM
+        bot_service = get_bot_services(user_profile_id)
```
I'm guessing we fetch that in this code path somewhere else; it would be best to not have to do it twice.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -535,35 +542,54 @@ def maybe_add_event(user_profile_id: int, bot_type: int) -> None:
+                )
+            case Service.TriggerType.SUBSCRIPTION:
+                if not is_stream:
+                    return
```
I don't think we need to filter this -- the intent is just for the bot to get outgoing webhooks like a normal user would. Maybe "SUBSCRIPTION` is the wrong name -- we could call it "RECIPIENT", perhaps, with the idea being that it'll be true if the user is a recipient?

PieterCK replied:
Ok, do we want the `evet["trigger']` value to be the same as the value of the trigger type? It looks to me that it's better to keep it as is:

The current `evet["trigger']` for each trigger types:
- "all_mentions" : "mentions"
- "dm_received" : "direct_message"
- "all_received" : "received_message" (?)

PieterCK commented on zerver/actions/message_send.py:
```diff
@@ -535,35 +542,54 @@ def maybe_add_event(user_profile_id: int, bot_type: int) -> None:
+                if not is_stream:
+                    return
+                bot_user_profile = get_user_profile_by_id(user_profile_id)
+                if not subscribed_to_stream(bot_user_profile, recipient_type_id):
```
Yep, it is.

PieterCK commented:
Still WIP, I'll post another comment once this is ready for review. I opened a discussion around the fields name and other details here: [#api design > Adding service bot trigger types. #34524](https://chat.zulip.org/#narrow/channel/378-api-design/topic/Adding.20service.20bot.20trigger.20types.2E.20.2334524/with/2169743)

PieterCK committed:
service_bot: Add a new service trigger, `all_received`.

This adds a new option for the Service.triggers field, `all_received`.
This trigger activates the bots Service whenever the bot receive any
message (part of the recipient).

PieterCK commented on zerver/views/users.py:
```diff
@@ -609,6 +633,9 @@ def add_bot_backend(
     except UserProfile.DoesNotExist:
         pass
 
+    if bot_type in UserProfile.SERVICE_BOT_TYPES and not service_triggers:
```
This check is here because just a few lines down from here we create the bots' user profile.

PieterCK commented on zerver/actions/message_send.py:
```diff
@@ -550,26 +584,62 @@ def maybe_add_event(user_profile_id: int, bot_type: int) -> None:
         if user_profile_id not in mentioned_user_ids and user_profile_id not in active_user_ids:
```
I think this condition is redundant because of the operation to create `relevant_service_bot_data`, I maybe we can drop this and move the note in the comment to `relevant_service_bot_data` ?

PieterCK committed:
migrations: Fix existing service bot with no triggers.

This adds a custom migration that updates existing Service bots triggers
field, setting them to ["all_mentions", "dm_received"] to maintain their
existing functions.

PieterCK commented on zerver/tests/test_service_bot_system.py:
```diff
@@ -621,9 +636,10 @@ def test_flag_messages_service_bots_has_processed(self) -> None:
-        message_id = self.send_group_direct_message(
-            sender, recipients, content=f"@**{self.bot_profile.full_name}** foo"
-        )
+        with self.assertLogs(level="INFO"):
```
This is because we don't patch `queue_event_on_commit` for this test:
```
Running zerver.tests.test_service_bot_system.TestServiceBotEventTriggers.test_flag_messages_service_bots_has_processed
2025-05-29 08:13:22.322 INFO [] Outgoing webhook request from 36@zulip took 0.001425 seconds
2025-05-29 08:13:22.329 INFO [] Outgoing webhook request from 35@zulip took 0.001472 seconds
2025-05-29 08:13:22.349 INFO [] Outgoing webhook request from 36@zulip took 0.001544 seconds
2025-05-29 08:13:22.353 ERR  [] Error: User 35 has bot with invalid embedded bot service foo
Running zerver.tests.test_service_bot_system.TestServiceBotEventTriggers.test_no_trigger_on_direct_message_from_bot_with_no_trigger
```

PieterCK commented on zerver/openapi/zulip.yaml:
```diff
@@ -24160,6 +24160,7 @@ components:
 
                       - "all_mentions" = Any mention of the service bot, regardless of whether
                         the bot is part of the recipient or not.
+                      - "all_received" = Whenever the bot receives any message.
```
hmm I wonder if this should just be included in the commit where we add the "triggers" section in the API doc, `message_send: Make service bot accept multiple triggers`?

PieterCK commented on zerver/lib/test_classes.py:
```diff
@@ -717,6 +718,15 @@ def notification_bot(self, realm: Realm) -> UserProfile:
+        ):
+            extras.update(
+                service_triggers=orjson.dumps(get_default_service_bot_triggers()).decode()
+            )
```
This is basically to maintain backwards compatibility for existing service bot related tests.

PieterCK commented on zerver/views/users.py:
```diff
@@ -499,7 +503,12 @@ def patch_bot_backend(
+        if len(service_triggers) < 1:
+            raise JsonableError(_("A service bot must have at least one trigger."))
+        do_update_outgoing_webhook_service(
+            bot, service_interface, service_payload_url, service_triggers
```
In practice, we always pass all the current values of the bots fields, when patching a field. So, this is done to avoid unnecessarily sending more events.

PieterCK commented:
@zulipbot add "buddy review"

PieterCK commented on zerver/models/bots.py:
```diff
@@ -49,6 +50,14 @@ class Service(models.Model):
+
+    message_trigger_type = models.CharField(
+        default=TriggerType.MENTION_ONLY, choices=TriggerType, max_length=1
+    )
```
Since we decided to support multiple triggers per bot, this is the current version:
```python
class Service(models.Model):
    ...
    BOT_TRIGGER_ALL_MENTIONS = "all_mentions"
    BOT_TRIGGER_DM_RECEIVED = "dm_received"
    BOT_TRIGGER_ALL_RECEIVED = "all_received"

    BOT_TRIGGER_CHOICES = [
        BOT_TRIGGER_ALL_MENTIONS,
        BOT_TRIGGER_DM_RECEIVED,
        BOT_TRIGGER_ALL_RECEIVED,
    ]

    triggers = JSONField(default=list)
```

PieterCK commented:
@Aditya8840 just checking in to see if you're still reviewing this one

PieterCK commented:
@zulipbot add "mentor review"

Aditya8840 commented:
> @Aditya8840 just checking in to see if you're still reviewing this one

Sorry for the delay — I got a bit overwhelmed by the code. I'll try to complete the review by today or tomorrow for sure.

PieterCK commented:
I pushed some changes to update the "Triggers" section in the `api/outgoing-webhook` doc:

| Changes |
|--------|
| (before) ![image](https://github.com/user-attachments/assets/a06b2a26-6052-40e8-a0cc-7418d05182f4)|
| (after) ![image](https://github.com/user-attachments/assets/3524096c-81ad-480e-8358-7937af7651bb)|

PieterCK commented:
@Aditya8840 I opened  #34990 and separated the first 4 prep commits to that PR to keep the changes at manageable size, would appreciate it if you could take a look at that first
